### ISSUE TITLE
Performance issues using powershell versions

### ISSUE THREAD

**Issue by csabo** [sensu-plugins/sensu-plugins-windows#60]
@majormoses requested I open this issue

We ran into performance issues using the powershell scripts for metrics and checks. As we started really growing in number of checks, and number of nodes.  Get-WMIObject isn't the fastest thing on the block.  I've rewritten the checks and metrics using CIM and the performance greatly increased. around 75% faster execution time, and lower CPU load during execution.

I've combined all of the metrics into a single script, because a single instantiation of powershell worked out better, and we gather all of this data from all windows nodes.

https://github.com/csabo/sensu_win_checks_cim

**Comment by majormoses** [comment]
@csabo Thanks for your hard work! I know next to nothing about powershell so I am not much help in actual porting but I am more than happy to review the PRs, run tests, (I have a windows desktop I use for gaming and do not have access to a windows server), help with changelog, etc.

Regarding the metrics scripts I think we would want to keep them separate as different people have different needs. If I knew what I was talking about in powershell I would say we could implement the logic in whatever they call a library and create individual scripts that just included it. If you wish to avoid the penalty cost of running each one as its own instantiation maybe we can still have a meta check to run all the checks but it should pull from the shared logic. I know in ruby and most interpreted languages that is easy to do without a performance penalty but I am not sure if that is possible here as I have not used windows for anything but playing games in many years.

**Comment by pdpvandyk** [comment]
I would be interested in helping to test or work it into a library.  I can't promise "timely" necessarily - but I've run into the same performance issues and I was going to tackle a rewrite, as well.

**Comment by majormoses** [comment]
@csabo that sounds good to me @pdpvandyk awesome no worries, this is open source and something like 99.99% of all plugin development is done by people (such as myself) who do not work for sensu and donate their time, expertise, etc  to better the sensu and larger monitoring community.

**Comment by csabo** [comment]
@pdpvandyk @majormoses I've gotten started on a branch that will include the CIM replacements of WMI where applicable.

**Comment by csabo** [comment]
@pdpvandyk @majormoses Alright the branch is up, Ive checked each with powershell 5 and 6, they seem good to me, a couple of the metrics I felt were a little complex in their code, so I replaced them with what I did for us in house, that I feel is a little more straight forward. The only ones I didnt get to yet were the "network" "uptime" and "system" metrics, I'll try to get them done tomorrow or this weekend. In the mean time plenty here to test/validate/discuss! 

https://github.com/csabo/sensu-plugins-windows/tree/cim
